src/v/cluster/plugin_frontend.h

```diff
@@ -0,0 +1,166 @@
+    // updates.
+    std::optional<model::transform_metadata>
+      lookup_transform(model::transform_id) const;
+
+    // Lookup transforms for input topics.
+    absl::flat_hash_map<model::transform_id, model::transform_metadata>
+      lookup_transforms_by_input_topic(model::topic_namespace_view) const;
+
+    // Get a snapshot of all the transforms that currently exist.
+    absl::flat_hash_map<model::transform_id, model::transform_metadata>
+    all_transforms() const;
+
+private:
+    // Perform a mutation request, check if this node is the cluster leader and
+    // ensuring that the right core is used for local mutations.
+    ss::future<mutation_result>
+      do_mutation(transform_cmd, model::timeout_clock::time_point);
+
+    // If this node is not the leader, dispatch this mutation to `node_id` as
+    // the cluster's leader.
+    ss::future<mutation_result> dispatch_mutation_to_remote(
+      model::node_id, transform_cmd, model::timeout_clock::duration);
+
+    // Performs (and validates) a mutation command to be inserted into the
+    // controller log.
+    //
+    // This must take place on the controller shard on the cluster leader to
+    // ensure consistency.
+    ss::future<mutation_result> do_local_mutation(
+      transform_cmd, model::offset, model::timeout_clock::time_point);
+
+    // If this command is a upsert, then the ID is assigned.
+    //
+    // The offset supplied is the latest offset applied in raft0
+    //
+    // This must take place on the controller shard on the cluster leader to
+    // ensure consistency.
+    void assign_id(transform_cmd*, model::offset);
+
+    // Ensures that the mutation is valid.
+    //
+    // This must take place on the controller shard on the cluster leader to
+    // ensure consistency.
+    errc validate_mutation(const transform_cmd&);
+
+public:
+    /**
+     * A validator of commands that come through, the validation must happen on
+     * the controller leader, core 0 so that these is a consistent view of the
+     * data in the plugin table.
+     *
+     * Mostly exposed seperately for unit testing.
+     */
+    class validator {
+    public:
+        validator(topic_table*, plugin_table*);
+
+        // Ensures that the mutation is valid.
+        //
+        // This must take place on the controller shard on the cluster leader to
+        // ensure consistency.
+        errc validate_mutation(const transform_cmd&);
+
+    private:
+        // Would adding this input->output transform cause a cycle?
+        bool would_cause_cycle(
+          model::topic_namespace_view, model::topic_namespace);
+
+        topic_table* _topics;
+        plugin_table* _table;
+    };
+
+private:
+    // Can only be accessed on the current shard.
+    model::node_id _self;
+    partition_leaders_table* _leaders;
+    rpc::connection_cache* _connections;
+    plugin_table* _table;
+    topic_table* _topics;
+    ss::abort_source* _abort_source;
```

Comment:
nit: the codebase generally uses references everywhere, may be good to stick to that for consistency (multiple places)

Reply:
Done. I raised a question in #cpplang about the clang-tidy warnings using references gives.